archi: init at 4.7.1

earldouglas opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change

Add a derivation for Archi.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

earldouglas committed:
Archi: init at 4.7.1

earldouglas commented on pkgs/tools/misc/Archi/default.nix:
```diff
@@ -0,0 +1,44 @@
+  installPhase = ''
+    cp -r . $out
+    mkdir -p $out/bin
+    ln -s $out/Archi $out/bin/Archi
```
This is creating an invalid relative symlink to `../Archi` under NixOS, but it works correctly on Nix+Linux.

SuperSandro2000 replied:
Can we use mv here?

earldouglas commented:
This isn't quite working yet; looks like the binary might need to be patched for NixOS paths.

SuperSandro2000 commented on pkgs/tools/misc/Archi/default.nix:
```diff
@@ -0,0 +1,44 @@
+{ stdenv, fetchurl }:
+
+let version = "4.7.1";
+in stdenv.mkDerivation {
+
+  pname = "Archi";
+  inherit version;
```
```suggestion
stdenv.mkDerivation rec {

  pname = "Archi";
  version = "4.7.1";
```

SuperSandro2000 commented on pkgs/tools/misc/Archi/default.nix:
```diff
@@ -0,0 +1,44 @@
+        url = "https://www.archimatetool.com/downloads/archi/Archi-Linux64-${version}.tgz";
+        sha256 = "0sd57cfnh5q2p17sd86c8wgmqyipg29rz6iaa5brq8mwn8ps2fdw";
+      }
+    else if stdenv.hostPlatform.system == "x86_64-linux" then
```
That is the same as the one above. Did you mean x86_64-darwin?

SuperSandro2000 commented on pkgs/tools/misc/Archi/default.nix:
```diff
@@ -0,0 +1,44 @@
+        sha256 = "0sd57cfnh5q2p17sd86c8wgmqyipg29rz6iaa5brq8mwn8ps2fdw";
+      }
+    else if stdenv.hostPlatform.system == "x86_64-linux" then
+      throw "Unsupported system"
```
I can test if this builds but you need to enable it.

SuperSandro2000 commented on pkgs/tools/misc/Archi/default.nix:
```diff
@@ -0,0 +1,44 @@
+    ln -s $out/Archi $out/bin/Archi
+  '';
+
+  meta = {
+    description = "ArchiMate modelling toolkit";
+    longDescription = ''
+      Archi is an open source modelling toolkit to create ArchiMate
+      models and sketches.
+    '';
+    homepage = "https://www.archimatetool.com/";
+    license = stdenv.lib.licenses.mit;
+    platforms = stdenv.lib.platforms.linux;
+    maintainers = [ stdenv.lib.maintainers.earldouglas ];
+  };
```
```suggestion
  meta = with stdenv.lib; {
    description = "ArchiMate modelling toolkit";
    longDescription = ''
      Archi is an open source modelling toolkit to create ArchiMate
      models and sketches.
    '';
    homepage = "https://www.archimatetool.com/";
    license = licenses.mit;
    platforms = platforms.linux;
    maintainers = with maintainers; [ earldouglas ];
  };
```

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -688,6 +688,8 @@ in
     inherit (darwin.apple_sdk.frameworks) Carbon Cocoa Kernel;
   };
 
+  Archi = callPackage ../tools/misc/Archi { };
```
```suggestion
  archi = callPackage ../tools/misc/archi { };
```

Names must be lowercased at the beginning.

earldouglas committed:
Update pkgs/tools/misc/Archi/default.nix

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

earldouglas committed:
Update pkgs/tools/misc/Archi/default.nix

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

earldouglas committed:
Update pkgs/top-level/all-packages.nix

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

SuperSandro2000 commented:
Darwin patch:
```patch
diff --git a/pkgs/tools/misc/Archi/default.nix b/pkgs/tools/misc/Archi/default.nix
index 137f679b1da0..fa2db41760a8 100644
--- a/pkgs/tools/misc/Archi/default.nix
+++ b/pkgs/tools/misc/Archi/default.nix
@@ -1,4 +1,4 @@
-{ stdenv, fetchurl }:
+{ stdenv, fetchurl, fetchzip }:
 
 stdenv.mkDerivation rec {
 
@@ -11,22 +11,21 @@ stdenv.mkDerivation rec {
         url = "https://www.archimatetool.com/downloads/archi/Archi-Linux64-${version}.tgz";
         sha256 = "0sd57cfnh5q2p17sd86c8wgmqyipg29rz6iaa5brq8mwn8ps2fdw";
       }
-    else if stdenv.hostPlatform.system == "x86_64-linux" then
-      throw "Unsupported system"
-      /*
-       * A Mac version is available; needs to be tested on a Mac system:
-       * fetchurl {
-       *   url = "https://www.archimatetool.com/downloads/archi/Archi-Mac-${version}.zip";
-       *   sha256 = "0zg0bs9i2kkr11bbssfxhrp4ym2zzhb2bk7m38z9cj99h7pmgyhr";
-       * }
-       */
+    else if stdenv.hostPlatform.system == "x86_64-darwin" then
+      fetchzip {
+        url = "https://www.archimatetool.com/downloads/archi/Archi-Mac-${version}.zip";
+        sha256 = "1h05lal5jnjwm30dbqvd6gisgrmf1an8xf34f01gs9pwqvqfvmxc";
+      }
     else
        throw "Unsupported system";
 
-  installPhase = ''
+  installPhase = if stdenv.hostPlatform.system == "x86_64-darwin" then ''
+    mkdir -p "$out/Applications"
+    mv Archi.app "$out/Applications/"
+  '' else ''
+    install -Dm755 Archi $out/bin/Archi
+    rm Archi
     cp -r . $out
-    mkdir -p $out/bin
-    ln -s $out/Archi $out/bin/Archi
   '';
 
   meta = with stdenv.lib; {
@@ -37,7 +36,7 @@ stdenv.mkDerivation rec {
     '';
     homepage = "https://www.archimatetool.com/";
     license = licenses.mit;
-    platforms = platforms.linux;
+    platforms = platforms.unix;
     maintainers = with maintainers; [ earldouglas ];
   };
 }
```

Also please squash the commits together.

earldouglas committed:
Incorporate Darwin patch by SuperSandro2000

earldouglas committed:
Rename Archi dir to archi

earldouglas committed:
Patch dependencies

This gets us a little closer, but we're not correctly including/patching
dependencies:

```
java.lang.UnsatisfiedLinkError: Could not load SWT library. Reasons:
        no swt-pi4-gtk-4932r18 in java.library.path
        no swt-pi4-gtk in java.library.path
        Can't load library: /home/james/.swt/lib/linux/x86_64/libswt-pi4-gtk-4932r18.so
        Can't load library: /home/james/.swt/lib/linux/x86_64/libswt-pi4-gtk.so
```

SuperSandro2000 commented on pkgs/tools/misc/archi/default.nix:
```diff
@@ -0,0 +1,64 @@
+}:
+
+stdenv.mkDerivation rec {
+
```
```suggestion
```

SuperSandro2000 commented on pkgs/tools/misc/archi/default.nix:
```diff
@@ -0,0 +1,64 @@
+  installPhase =
+    if stdenv.hostPlatform.system == "x86_64-linux" then
+      ''
+        for i in configuration features p2 plugins Archi.ini Archi
+        do
```
```suggestion
        for f in configuration features p2 plugins Archi.ini Archi; do
```

SuperSandro2000 commented on pkgs/tools/misc/archi/default.nix:
```diff
@@ -0,0 +1,64 @@
+      ''
+        for i in configuration features p2 plugins Archi.ini Archi
+        do
+          mkdir -p $out/bin
```
```suggestion
```

SuperSandro2000 commented on pkgs/tools/misc/archi/default.nix:
```diff
@@ -0,0 +1,64 @@
+
+  installPhase =
+    if stdenv.hostPlatform.system == "x86_64-linux" then
+      ''
```
```suggestion
      ''
      mkdir -p $out/bin
```

SuperSandro2000 commented on pkgs/tools/misc/archi/default.nix:
```diff
@@ -0,0 +1,64 @@
+        for i in configuration features p2 plugins Archi.ini Archi
+        do
+          mkdir -p $out/bin
+          cp -rf $i $out/bin/
```
```suggestion
          cp $f $out/bin/
```

SuperSandro2000 commented on pkgs/tools/misc/archi/default.nix:
```diff
@@ -0,0 +1,64 @@
+        done
+
+        install -D -m755 Archi $out/bin/Archi
+        rm Archi
```
```suggestion
```

SuperSandro2000 commented on pkgs/tools/misc/archi/default.nix:
```diff
@@ -0,0 +1,64 @@
+    '';
+    homepage = "https://www.archimatetool.com/";
+    license = licenses.mit;
+    platforms = platforms.linux;
```
```suggestion
    platforms = platforms.linux ++ platforms.darwin;
```

SuperSandro2000 commented:
Result of `nixpkgs-review pr 103355` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>archi</li>
  </ul>
</details>

earldouglas committed:
Apply suggestions from code review

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

SuperSandro2000 commented:
Result of `nixpkgs-review pr 103355` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>archi</li>
  </ul>
</details>
